pkg/pillar/pubsub/socketdriver/publish.go

```diff
@@ -308,3 +326,40 @@ func (s *Publisher) sendComplete(sock net.Conn) error {
 	_, err := sock.Write([]byte(buf))
 	return err
 }
+
+// check and waits till conn's fd is readable
+func connReadCheck(conn net.Conn) error {
+	var sysErr error
+
+	sysConn, ok := conn.(syscall.Conn)
+	if !ok {
+		return nil
+	}
+	rawConn, err := sysConn.SyscallConn()
+	if err != nil {
+		return err
+	}
+
+	err = rawConn.Read(func(fd uintptr) bool {
```

Comment:
As we discussed, can use a RecvFrom with a zero length buffer to wait for readable.
That should handle the error cases and will not block the thread in the read system call.